### ISSUE TITLE
Loading PR list blocks await JoinableTaskFactory.SwitchToMainThreadAsync

### ISSUE THREAD

**Issue by jcansdale** [github/VisualStudio#1537]
#### How to repro

1. Install the following repro
[AsyncPackageRepro.zip](https://github.com/github/VisualStudio/files/1815100/AsyncPackageRepro.zip)
```
    [Guid("aae76547-10c6-4a2d-b33a-76ded02ac07b")]
    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
    [ProvideAutoLoad(Constants.vsContextNoSolution, PackageAutoLoadFlags.BackgroundLoad)]
    public sealed class MyAsyncPackage : AsyncPackage
    {
        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
        {
            Trace.WriteLine("Executed when VS opens without a solution");
            await JoinableTaskFactory.SwitchToMainThreadAsync();
            Trace.WriteLine("Executed when PR list on GitHub pane has finished loading");
        }
    }
```

2. Open Visual Studio 2015 with the `GitHub` pane visible and the PR list loading
3. The `await JoinableTaskFactory.SwitchToMainThreadAsync()` line only returns once the PR list has completely finished refreshing (which can take a long time)

#### What is impacted

- This is a problem when installing the PR status bar UI (which must be done on the Main thread). The current PR sometimes don't appear for a long time.
- This will also be a problem for packages that auto-load in order to handle command visibility (e.g. `GitHubPackage` and `InlineReviewsPackage`).
- Any other extensions that use `SwitchToMainThreadAsync` when they auto-load

#### Notes

This hasn't been a problem in the past because we haven't been using the `AllowsBackgroundLoading = true` and `PackageAutoLoadFlags.BackgroundLoad` options. Developers are now being strongly encouraged to enable these options. We will potentially be delaying when other extensions wake up.

This isn't necessarily a problem when a command is executed because in this case `InitializeAsync` starts off on the Main thread and `SwitchToMainThreadAsync` is a nop. If however auto-load was started before the command is executed, it will likely still get blocked.

#### How to fix

This can be resolved from inside `InitializeAsync` by using the following:
```
// `JoinableTaskFactory` is a property on `AsyncPackage`
await JoinableTaskFactory
   .WithPriority(VsTaskRunContext.UIThreadNormalPriority)
   .SwitchToMainThreadAsync();
```

We should also consider refreshing the PR list at a lower priority in order not to block other extensions.

#### Related 

- Fix deadlock when VS is opended with GitHub pane visible: #1536

**Comment by AArnott** [comment]
> Developers are now being strongly encouraged to enable these options. We will potentially be delaying when other extensions wake up.

Yes, that's actually the idea. Customers see a wide variance in startup performance and we're taking steps to try to ensure that startup and solution load happen as fast as possible since that tends to be what they care about most.

For a workaround, as I mention on gitter.im:

```cs
await this.JoinableTaskFactory.RunAsync(
    VsTaskRunContext.UIThreadNormalPriority,
    async delegate {
       await this.JoinableTaskFactory.SwitchToMainThreadAsync();
       // away you go
    });
```

Keep in mind though that you should use this sparingly, and do as little work as possible within such a pre-empting delegate as this.

**Comment by AArnott** [comment]
> Unfortunately the following doesn't seem to work. SwitchToMainThreadAsync still doesn't return until after the GitHub pane has finished refreshing (this can take a while).

I'm not sure why the GitHub pane's refresh step would block a request to switch to the main thread unless that pane refresh was blocking the UI thread with its work. If that's the case, can it do some of this work either asynchronously or off the UI thread?

**Comment by sharwell** [comment]
@AArnott The GitHub extension uses Reactive-UI instead of vs-threading. It also doesn't appear to reset the UI scheduler priority (which [defaults to `DispatcherPriority.Normal`](https://github.com/Reactive-Extensions/Rx.NET/blob/ba98e6254c9a2f841cbc4169bf38590b133c8064/Rx.NET/Source/src/System.Reactive/Platforms/Windows/Concurrency/DispatcherScheduler.cs#L51)), so until this is corrected the GitHub extension UI has ultimate priority when running in VS.

**Comment by AArnott** [comment]
Ya, anything with Normal priority will starve out not only background work but even user input as well.
That would be a good thing for github's use of Rx to change.

**Comment by jcansdale** [comment]
@AArnott @sharwell,

Thanks for all the info! We definitely want to fix this issue at source rather than use workarounds. Luckily it only seems to be an issue when VS opens with the GitHub pane active.

**Comment by jcansdale** [comment]
@AArnott this seems to be working well. 😄 

I was wondering if there might be a simpler form of this if the code I'm executing isn't async?

For example, I'm currently doing this:

```cs
await JoinableTaskFactory.RunAsync(VsTaskRunContext.UIThreadNormalPriority, async () =>
{
    await JoinableTaskFactory.SwitchToMainThreadAsync();    
    menuService.AddCommands(....); // non-async code
});
```

The following doesn't compile, but I was wondering if something along these lines might?
```cs
await JoinableTaskFactory.Run(
    VsTaskRunContext.UIThreadNormalPriority,
    () => menuService.AddCommands(....));
```

Basically an equivalent of `Task.Run(....)`, but running at `UIThreadNormalPriority`?

**Comment by AArnott** [comment]
@jcansdale What is your objection to using async? There's no way to avoid it. If you were already on the UI thread, you wouldn't have this issue to begin with. Switching to the UI thread is an async operation -- you shouldn't block the calling thread to do it.

**Comment by sharwell** [comment]
> Switching to the UI thread is an async operation -- you shouldn't block the calling thread to do it.

:memo: Keep in mind this isn't simply a point of "A is better than B because it's aligns with some mental coding model that I think is best". We regularly file internal bugs where blocking operations on a background thread (particularly thread pool threads) cause performance problems because they tried to synchronously invoke an operation on the UI thread.

**Comment by AArnott** [comment]
True. The real point though is that `JoinableTaskFactory` *never* switches to the UI thread at all, without an async delegate that includes `await JTF.SwitchToMainThreadAsync();` in it. That's true even if you pass in a UI thread priority as a first argument. That first arg only influences what the UI priority will be when you later request a switch -- it doesn't do so implicitly.

**Comment by jcansdale** [comment]
@AArnott,

> What is your objection to using async?

No particular objection to using async, it's just the methods that need to be the Main thread tend not to be async. It tends to be create a few commands using MEF and add them to `IMenuCommandService`.  For example:

https://github.com/github/VisualStudio/blob/7d4c45f3de7f0de412d272a6f33426dd4109e00b/src/GitHub.InlineReviews/InlineReviewsPackage.cs#L36

I was just curious if there was a simpler way to execute an action on the Main thread that avoids the risk of deadlock. 😉

**Comment by jcansdale** [comment]
> That first arg only influences what the UI priority will be when you later request a switch -- it doesn't do so implicitly.

Ah, in that case the following seems to make sense. I wasn't 100% sure if `RunAsync` would change anything immediately:

```
await JoinableTaskFactory.RunAsync(VsTaskRunContext.UIThreadNormalPriority, InitializeMenus);
```

**Comment by AArnott** [comment]
Ya, it's an unfortunate artifact of the collision of APIs and extension methods. The JTF.RunAsync method you're calling is actually an extension method provided by VS to add VS-specific priorities. We had a choice of whether to automatically switch to the UI thread for the delegate you pass in, since the first arg would seem to suggest that might happen, or to hold to the built-in JTF.RunAsync methods which never switch contexts before invoking the delegate. There were pros/cons to each. 
For better or worse, we went with the policy that all JTF.RunAsync overloads (including extension methods) should have the same behavior when invoking the delegate, which is to invoke it immediately on the calling thread.
